File: backend/crates/domain/coffret-format/src/control/decode.rs
use coffret_model::ControlObjectName;

use super::ceiling::check_control_object_len;
use super::decoded_object::DecodedControlObject;
use super::header::ControlHeader;
use super::payload;
use crate::aead::Cipher;
use crate::error::{Error, Result};
use crate::purpose::Purpose;
use crate::purpose_key::PurposeKey;

/// Opens a control object stored under `object_name`.
///
/// The name is part of what is checked, not decoration: recovery finds these
/// objects by name, so a name that does not lead to the object it promised is a
/// disagreement about what the object *is* and the object is rejected. The kind
/// is checked against FM-12's admission table rather than for equality, because
/// one name form covers the whole control-head chain — `head-<generation>`
/// admits an ordinary Journal record and the Index Snapshot that activates an
/// epoch, and nothing else. The generation and the replica position are the
/// name's alone to state, so those are checked for equality. All of it is on
/// plaintext bytes, before the key is used at all.
///
/// The object's own size is checked there too, against the ceiling its kind
/// carries. A caller that read the object off Storage was already bounded by the
/// same ceiling before it buffered anything; repeating it here is what keeps the
/// guarantee a property of the format rather than of one call site, and it costs
/// a comparison on bytes already in hand.
pub fn decode_control_object(
    object: &[u8],
    object_name: &str,
    key: &PurposeKey,
) -> Result<DecodedControlObject> {
    let name = ControlObjectName::parse(object_name)?;
    let header = ControlHeader::parse(object)?;
    if !name.admits(header.kind) {
        return Err(Error::ControlObjectKindNotAdmitted {
            name,
            kind: header.kind,
        });
    }
    if name.generation() != header.generation {
        return Err(Error::ObjectNameMismatch {
            field: "generation",
        });
    }
    if name.replica() != header.replica {
        return Err(Error::ObjectNameMismatch {
            field: "replica position",
        });
    }
    check_control_object_len(header.kind, object.len() as u64)?;

    let key = key.require(Purpose::of_control_object(header.kind))?;
    let (associated_data, message) = object.split_at(ControlHeader::LEN);
    if message.is_empty() {
        return Err(Error::MissingControlPayload);
    }

    // The associated data is the header exactly as it appears in the object.
    let plaintext = Cipher::new(key).open(header.nonce(), associated_data, message)?;
    Ok(DecodedControlObject {
        kind: header.kind,
        generation: header.generation,
        replica: header.replica,
        payload: payload::decode(&plaintext)?,
    })
}

File: frontend/packages/domain/format/src/control/decode.ts
import { open } from '../internal/aead.js';
import { fail } from '../errors.js';
import { purposeKeyBytes, purposeOfControlObject, type PurposeKey } from '../purposeKey.js';
import { CONTROL_HEADER_LENGTH, parseControlHeader } from './header.js';
import { decodeControlPayload, type ControlPayload } from './payload.js';
import { nameAdmitsKind, parseControlObjectName } from './objectName.js';
import type { Generation } from '../model/generation.js';
import type { ReplicaPosition } from '../model/replicaPosition.js';
import type { ControlObjectKind } from '../model/kinds.js';

/** An opened control object: what its header said, and what it carried. */
export interface DecodedControlObject {
  /** Which kind of control state this object carries. */
  kind: ControlObjectKind;
  /** Where the object sat in the Library's control history (FM-13). */
  generation: Generation;
  /** Which replica this is, out of how many. */
  replica: ReplicaPosition;
  /** The payload, with the epoch that wrote it. */
  payload: ControlPayload;
}

/**
 * Opens a control object stored under `objectName`.
 *
 * The name is part of what is checked, not decoration: recovery discovers these
 * objects by name, while what an object *is* rides in its authenticated header.
 * A name that did not lead to the object it promised is therefore a
 * disagreement about what the object is, and the object is rejected (FM-12).
 * The kind is checked against FM-12's admission table rather than for equality,
 * because one name form covers the whole control-head chain —
 * `head-<generation>` admits an ordinary Journal record and the Index Snapshot
 * that activates an epoch, and nothing else. The generation and the replica
 * position are the name's alone to state, so those are checked for equality.
 * All of it is on plaintext bytes, before the key is used at all.
 */
export function decodeControlObject(
  object: Uint8Array,
  objectName: string,
  key: PurposeKey,
): DecodedControlObject {
  const name = parseControlObjectName(objectName);
  const header = parseControlHeader(object);
  if (!nameAdmitsKind(name, header.kind)) {
    fail(
      'control_object_kind_not_admitted',
      `${JSON.stringify(objectName)} admits no control object of kind ${header.kind}`,
    );
  }
  if (!name.generation.equals(header.generation)) {
    fail('object_name_mismatch', 'the object name and its header disagree on generation');
  }
  if (!name.replica.equals(header.replica)) {
    fail('object_name_mismatch', 'the object name and its header disagree on replica position');
  }

  const keyBytes = purposeKeyBytes(key, purposeOfControlObject(header.kind));
  const associatedData = object.subarray(0, CONTROL_HEADER_LENGTH);
  const message = object.subarray(CONTROL_HEADER_LENGTH);
  if (message.length === 0) {
    fail('missing_control_payload', 'control object carries no payload');
  }

  // The associated data is the header exactly as it appears in the object.
  const plaintext = open(keyBytes, header.nonce, associatedData, message);
  return {
    kind: header.kind,
    generation: header.generation,
    replica: header.replica,
    payload: decodeControlPayload(plaintext),
  };
}
